x/auth/vesting/msg_server.go

```diff
@@ -99,6 +99,92 @@ func (s msgServer) CreateVestingAccount(goCtx context.Context, msg *types.MsgCre
 	return &types.MsgCreateVestingAccountResponse{}, nil
 }
 
+func min64(i, j int64) int64 {
+	if i < j {
+		return i
+	}
+	return j
+}
+
+// mergePeriods returns the merge of two vesting period schedules.
+// The merge is defined as the union of the vesting events, with simultaneous
+// events combined into a single event.
+// Returns new start time, new end time, and merged vesting events, relative to
+// the new start time.
+func mergePeriods(startP, startQ int64, p, q []types.Period) (int64, int64, []types.Period) {
+	timeP := startP                    // time of last merged p event, next p event is relative to this time
+	timeQ := startQ                    // time of last merged q event, next q event is relative to this time
+	startTime := min64(startP, startQ) // we pick the earlier time
+	time := startTime                  // time of last merged event, or the start time
+	iP := 0                            // p indexes before this have been merged
+	iQ := 0                            // q indexes before this have been merged
+	lenP := len(p)
+	lenQ := len(q)
+	merged := []types.Period{}
+	for iP < lenP && iQ < lenQ {
+		nextP := timeP + p[iP].Length // next p event in absolute time
+		nextQ := timeQ + q[iQ].Length // next q event in absolute time
+		if nextP < nextQ {
+			// consume from p
+			period := types.Period{
+				Length: nextP - time,
+				Amount: p[iP].Amount,
+			}
+			merged = append(merged, period)
+			timeP = nextP
+			time = nextP
+			iP++
```

Comment:
Could you please factor this logic out into a `consume` function?  It could be either a local-to-mergePeriods closure to avoid passing common arguments, or a file-level helper function.

That would help keep things DRY, make it easier for me to review, and avoid copy-pasta errors.

Reply:
I've refactored along the lines you suggested and I think it's a net improvement.  It still uses a grab-bag of function-local variables to hold the iteration state, but I think data abstraction would be more trouble than it's worth.

Another approach I considered was to convert the vesting schedules to use absolute time, merge the schedules more simply, then convert back to relative times. Again, I don't think it's worth the effort.